tools/conformance/djxl_conformance.cc

```diff
@@ -236,16 +233,79 @@ class JSONArray : public JSONField {
 
+bool DecodeJXLToJpeg(const char* input_path, const char* output_path) {
+  // JPEG output buffer when reconstructing a JPEG file.
+  std::vector<uint8_t> jpeg_data;
+  std::vector<uint8_t> jpeg_data_chunk(16 * 1024);
+  auto dec = JxlDecoderMake(nullptr);
+
+  uint32_t events = JXL_DEC_JPEG_RECONSTRUCTION;
+  EXPECT_SUCCESS(JxlDecoderSubscribeEvents(dec.get(), events));
+
+  // TODO(deymo): Consider using a multi-threading decoder for conformance
+  // testing as well.
+
+  // Load and set input all at oncee.
+  std::vector<uint8_t> jxl_input;
+  EXPECT_TRUE(LoadFile(input_path, &jxl_input));
+  EXPECT_SUCCESS(
+      JxlDecoderSetInput(dec.get(), jxl_input.data(), jxl_input.size()));
+
+  bool has_jpeg_reconstruction = false;
+
+  while (true) {
+    JxlDecoderStatus status = JxlDecoderProcessInput(dec.get());
+    if (status == JXL_DEC_ERROR) {
+      fprintf(stderr, "Error decoding.\n");
+      return false;
+    } else if (status == JXL_DEC_NEED_MORE_INPUT) {
+      fprintf(stderr, "Error decoding: expected more input.\n");
+      return false;
+    } else if (status == JXL_DEC_JPEG_RECONSTRUCTION) {
+      has_jpeg_reconstruction = true;
+      // Decoding to JPEG.
+      EXPECT_SUCCESS(JxlDecoderSetJPEGBuffer(dec.get(), jpeg_data_chunk.data(),
+                                             jpeg_data_chunk.size()));
+    } else if (status == JXL_DEC_JPEG_NEED_MORE_OUTPUT) {
+      // Decoded a chunk to JPEG.
+      size_t used_jpeg_output =
+          jpeg_data_chunk.size() - JxlDecoderReleaseJPEGBuffer(dec.get());
+      jpeg_data.insert(jpeg_data.end(), jpeg_data_chunk.data(),
+                       jpeg_data_chunk.data() + used_jpeg_output);
+      EXPECT_SUCCESS(JxlDecoderSetJPEGBuffer(dec.get(), jpeg_data_chunk.data(),
+                                             jpeg_data_chunk.size()));
+    } else if (status == JXL_DEC_SUCCESS) {
+      break;
+    } else {
+      fprintf(stderr, "Error: unexpected status: %d\n",
+              static_cast<int>(status));
+      return false;
+    }
+  }
+
+  if (has_jpeg_reconstruction) {
+    size_t used_jpeg_output =
+        jpeg_data_chunk.size() - JxlDecoderReleaseJPEGBuffer(dec.get());
+    jpeg_data.insert(jpeg_data.end(), jpeg_data_chunk.data(),
+                     jpeg_data_chunk.data() + used_jpeg_output);
+    EXPECT_TRUE(SaveFile(output_path, jpeg_data));
+  }
+  return true;
+}
+
 bool DecodeJXL(const DecodeOptions& opts) {
   auto dec = JxlDecoderMake(nullptr);
 
-  uint32_t events = JXL_DEC_BASIC_INFO | JXL_DEC_COLOR_ENCODING;
+  uint32_t events =
+      JXL_DEC_BASIC_INFO | JXL_DEC_COLOR_ENCODING | JXL_DEC_PREVIEW_IMAGE;
   if (opts.pixel_prefix) events |= JXL_DEC_FRAME | JXL_DEC_FULL_IMAGE;
-  if (opts.jpeg_path) events |= JXL_DEC_JPEG_RECONSTRUCTION;
-  if (opts.preview) events |= JXL_DEC_PREVIEW_IMAGE;
   // We need to output the frame header info in the metadata.
   if (opts.metadata_path) events |= JXL_DEC_FRAME;
 
+  if (opts.jpeg_path) {
```

Comment:
Why not do `events = JXL_DEC_JPEG_RECONSTRUCTION` in this case if this is what you want? (if jpeg_path then nothing else) I think it should be possible to get both from the API.

Reply:
@lvandeve AFAIU you cannot get both JPEG and pixels, right?

Reply:
I was proposing to rollback the part that creates a new function with a new decoder and decoder loop and instead set `events = JXL_DEC_JPEG_RECONSTRUCTION` here. You'd need to handle whether jpeg is written after the loop. That's not the same as getting both JPEG and pixels (which should be allowed IMHO but that's a separate question, at least basic info should be allowed).

Reply:
I'm not sure this would be helpful - `djxl_conformance` needs to output both JPEG and pixels anyway, are you suggesting we use a single function that will do one or the other and then change the set of options passed to it?